index.bs

```diff
@@ -2772,6 +2865,102 @@ The [=remote end steps=] with |session| and |command parameters| are:
 
 </div>
 
+## The browser Module ## {#module-browser}
+
+The <dfn export for=modules>browser</dfn> module contains commands for
+managing the remote end browser process.
+
+### Definition ### {#module-browser-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+BrowserCommand = (
+  browser.Close
+)
+</pre>
+
+[=local end definition=]
+
+<!-- Nothing yet -->
+<pre class="cddl local-cddl">
+</pre>
+
+
+### Commands ### {#module-browser-commands}
+
+#### The browser.close Command #### {#command-browser-close}
+
+The <dfn export for=commands>browser.close</dfn> command terminates all
+WebDriver sessions and cleans up automation state in the remote browser instance.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+      <pre class="cddl remote-cddl">
+      browser.Close = {
+        method: "browser.close",
+        params: EmptyParams,
+      }
+      </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+      <pre class="cddl">
+      EmptyResult
+      </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for browser.close">
+The [=remote end steps=] with |session| and <var ignore>command parameters</var> are:
+
+1. [=End the session=] with |session|.
+
+1. If [=active sessions=] is not [=list/empty=] an implementation may
+   return [=error=] with [=error code=] [=unable to close browser=], and then
+   run the following steps [=in parallel=]:
+
+    1. Wait until the [=Send a WebSocket message=] steps have been called with the
+       response to this command.
+
+    1. [=Cleanup the session=] with |session|.
+
+   Note: The behaviour in cases where the browser has multiple automation
+   sessions is currently unspecified. It might be that any session can close the
+   browser, or that only the final open session can actually close the browser,
+   or only the first session started can. This behaviour might be fully
+   specified in a future version of this specification.
+
+1. For each |active session| in [=active sessions=]:
+
+  1. [=End the session=] |active session|.
+
+  1. [=Cleanup the session=] with |active session|
+
+1. Return [=success=] with data null, and run the following steps [=in parallel=].
+
+  1. Wait until the [=Send a WebSocket message=] steps have been called with the
+     response to this command.
```

Comment:
What kind of response do you actually expect to see here for the other open sessions that did not trigger the shutdown? Note that the `id` of the response will not match with any command as sent such a client. Maybe we should send an event like `browser.closing` instead?

Reply:
None. I  just expect their WebSocket connections to close. Since clients have to cope with the websocket connection closing at an arbitrary time, a clean close at the websocket protocol level seems like enough of a signal that something closed the browser. It's of course also possible that something just sends it SIGKILL and we get an unclean shutdown.

Given that, I'm not sure an extra event adds much, but we could always add one later if needed. The only reason to respond to the command at all really is to ensure that we can provide an error in the case that shutdown is blocked.

Reply:
Yes, adding the event at a later time sounds good. I was only worried about other clients being able to distinct between a clean and expected shutdown, or a browser crash or WebSocket disconnect.